### PR TITLE
feat: Add AnchorNavigation component

### PR DESCRIPTION
**Pull request by rubencarvalho**
### Description

Adding the new Anchor Navigation component.

### How has this been tested?

Unit & integration tests. 

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/anchor-navigation/interfaces.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  /**
+   * List of anchors.
+   */
+  anchors: AnchorNavigationProps.Anchor[];
+
+  /**
+   * Specifies the active anchor href.
+   * For using the component in a controlled manner, use together with 'disableTracking'.
+   */
+  activehref?: string;
```
Comment (jperals) [inline src/anchor-navigation/interfaces.ts:26]: Is the "h" intentionally lowercase?

The Side navigation component has `activeHref` with uppercase "H".

Comment (rubencarvalho) [reply src/anchor-navigation/interfaces.ts]: Not intentional! Fixed it.

**Comment by jperals** [comment]
I think it would be good to add integration tests for features like scrolling on click and the scroll spy, as these are critical features which otherwise are not covered.

File: src/anchor-navigation/internal.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    setDisableTracking(disableTracking);
+
+    if (activeHref) {
+      setCurrentActiveHref(activeHref);
+    }
+  }, [setDisableTracking, setCurrentActiveHref, disableTracking, activeHref]);
+
+  useEffect(() => {
+    if (currentActiveHref) {
+      onActiveAnchorChangeHandler(memoizedAnchors.find(anchor => anchor.href === currentActiveHref));
```
Comment (jperals) [inline src/anchor-navigation/internal.tsx:52]: It would be nice to have a test that asserts that `onActiveAnchorChange` is called once and only once on anchor change.

File: src/anchor-navigation/internal.tsx [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  onActiveAnchorChange,
+  disableTracking = false,
+  activeHref = '',
+  __internalRootRef = null,
+  ...props
+}: AnchorNavigationProps & InternalBaseComponentProps) {
+  const baseProps = getBaseProps(props);
+
+  const hrefs = useMemo(() => anchors.map(anchor => anchor.href), [anchors]);
+  const memoizedAnchors = useMemo(() => anchors, [anchors]);
```
Comment (jperals) [inline src/anchor-navigation/internal.tsx:27]: What is this line for?

File: src/anchor-navigation/internal.tsx [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+          );
+        })}
+      </ol>
+    </nav>
+  );
+}
+
+interface AnchorProps {
+  anchor: AnchorNavigationProps.Anchor;
+  fireFollow: (anchor: AnchorNavigationProps.Anchor, event: React.SyntheticEvent | Event) => void;
```
Comment (jperals) [inline src/anchor-navigation/internal.tsx:82]: Why `fireFollow` and not `onFollow`?

File: src/anchor-navigation/scroll-spy.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  hrefs,
+}: {
+  hrefs: string[];
+}): [
+  string | undefined,
+  React.Dispatch<React.SetStateAction<string | undefined>>,
+  React.Dispatch<React.SetStateAction<boolean>>
+] {
+  const [currentHref, setCurrentHref] = useState<string>();
+  const [scrollY, setScrollY] = useState(isBrowser ? window.pageYOffset : 0);
```
Comment (jperals) [inline src/anchor-navigation/scroll-spy.tsx:17]: Because `scrollY` is a state variable, every time the user scrolls the component will re-render, even if it does not need to. Is there a way we can optimize this?

File: src/anchor-navigation/scroll-spy.tsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  const [disableTracking, setDisableTracking] = useState(false);
```
Comment (jperals) [inline src/anchor-navigation/scroll-spy.tsx:18]: Could we pass `disableTracking` directly as a prop so we don't need this extra state variable? This seems unnecessary duplication.

File: src/anchor-navigation/scroll-spy.tsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    }
+  }, [updateScroll, scrollY]);
+
+  useEffect(() => {
+    if (disableTracking || !isBrowser) {
+      return;
+    }
+
+    // Main logic to get the new active href
+    let newCurrentHref;
```
Comment (jperals) [inline src/anchor-navigation/scroll-spy.tsx:81]: Minor: no need to use a `let` and reassign, you can just use `setNewCurrentRef` twice.

File: src/anchor-navigation/scroll-spy.tsx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    const lastAnchorId = hrefs[hrefs.length - 1]?.slice(1);
+    const lastAnchorElementExists = document.getElementById(lastAnchorId);
+    return !!lastAnchorElementExists && scrollY >= Math.floor(document.body.scrollHeight - window.innerHeight);
+  }, [scrollY, hrefs]);
+
+  // Find the href for which the element is within the viewport plus EXTRA_OFFSET
+  const findHrefInView = useCallback(() => {
+    return hrefs.find(href => {
+      const rect = getRectByHref(href);
+      return rect && rect.bottom <= window.innerHeight && rect.bottom - EXTRA_OFFSET >= 0;
```
Comment (jperals) [inline src/anchor-navigation/scroll-spy.tsx:47]: I've noticed that when clicking on an anchor for which the content is shorter than the viewport, the next section will be activated instead (you can try this on the examples you provide in this PR using a big monitor: clicking e.g, on section 1.1 will highlight section 1.1.1).

Note that `rect.bottom` returns the coordinate _from the top of the viewport_ (not from the bottom) of the element's bottom edge, therefore `rect.bottom - EXTRA_OFFSET` will be negative for the element that we scrolled to the top. This might be leading to this function not returning any element and then falling back to `findLastHrefInView` as per the logic in line 86 below.

Comment (rubencarvalho) [reply src/anchor-navigation/scroll-spy.tsx]: I've noticed testing. `EXTRA_OFFSET` is now the `scrollSpyOffset`, which should fix this.
I'll make sure to add integration tests covering these edge cases.

File: src/anchor-navigation/scroll-spy.tsx [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+      window.addEventListener('scroll', updateScroll, {
+        capture: false,
+        passive: true,
+      });
+
+      return () => {
+        window.removeEventListener('scroll', updateScroll);
+      };
+    }
+  }, [updateScroll, scrollY]);
```
[the remark this answers is not in the document]
Comment (rubencarvalho) [reply src/anchor-navigation/scroll-spy.tsx]: No, cleaned up!

File: src/anchor-navigation/internal.tsx (hunk 4 above)
Comment (rubencarvalho) [reply src/anchor-navigation/internal.tsx:83]: Was trying to follow the convention used in SideNavigation, but it was also confusing to me so I renamed it! 👍

File: src/anchor-navigation/internal.tsx (hunk 2 above)
Comment (rubencarvalho) [reply src/anchor-navigation/internal.tsx]: TODO

File: pages/anchor-navigation/expandable.page.tsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+      <ScreenshotArea>
+        <SpaceBetween size="l">
+          <div className={styles['content-grid']}>
+            <TextContent />
+            <div>
+              <div className={styles['anchor-navigation']}>
+                <ExpandableSection variant="navigation" headerText="On this page">
+                  <AnchorNavigation
+                    ariaLabelledby="anchor-nav-heading"
+                    anchors={[
```
[the remark this answers is not in the document]
Comment (rubencarvalho) [reply pages/anchor-navigation/expandable.page.tsx:57]: Added

File: src/anchor-navigation/scroll-spy.tsx [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+      setCurrentHref(hrefs[hrefs.length - 1]);
+    } else {
+      setCurrentHref(findHrefInView() || (scrollY > 0 ? findLastHrefInView() : undefined));
+    }
+  }, [disableScrollSpy, isPageBottom, findHrefInView, findLastHrefInView, hrefs]);
+
+  useEffect(() => {
+    if (isBrowser) {
+      handleScroll();
+      window.addEventListener('scroll', handleScroll, { passive: true });
```
Comment (jperals) [inline src/anchor-navigation/scroll-spy.tsx:69]: Did you consider using the Intersection Observer API instead? https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API

Comment (rubencarvalho) [reply src/anchor-navigation/scroll-spy.tsx:69]: I did consider it initially. In fact, I implemented it (https://github.com/cloudscape-design/components/pull/1396/commits/00a711b3bcaf3421f1e6fdcc1657e3c503ca1422) but ran into some issues that made the code more complex than I would've liked...

Comment (rubencarvalho) [reply src/anchor-navigation/scroll-spy.tsx:69]: There were some edge cases (scrolled all the way to the bottom, in which the element didn't intersect the IO), lack of granularity, IO sometimes not being triggered (when being linked to a specific section of the page or quickly scrolling on the trackpad)

File: pages/anchor-navigation/expandable.page.tsx [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        <TextSample />
+        <h5 id="section-1-2-1-1">Section 1.2.1.1</h5>
+        <TextSample />
+      </div>
+    </SpaceBetween>
+  );
+};
+
+export default function ExpandableAnchorNavigation() {
+  const navigateToItem = (id: string) => {
```
Comment (jperals) [inline pages/anchor-navigation/expandable.page.tsx:35]: Nitpick: this function could be in `utils.tsx` so it can be de-duplicated

File: src/anchor-navigation/internal.tsx [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+  const hrefs = useMemo(() => anchors.map(anchor => anchor.href), [anchors]);
+
+  const onFollowHandler = useCallback(
+    (anchor: AnchorNavigationProps.Anchor, sourceEvent: React.SyntheticEvent | Event) => {
+      fireCancelableEvent(onFollow, anchor, sourceEvent);
+    },
+    [onFollow]
+  );
+
+  const onActiveHrefChangeHandler = useStableCallback((newActiveAnchor: AnchorNavigationProps.Anchor | undefined) => {
```
Comment (jperals) [inline src/anchor-navigation/internal.tsx:35]: Is this function necessary? Could we just call `fireNonCancellableEvent` directly on line 47 below instead?

File: src/anchor-navigation/scroll-spy.tsx [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+  // Check if we're scrolled to the bottom of the page
+  const isPageBottom = useCallback(() => {
+    return lastAnchorElementExists && window.scrollY >= Math.floor(document.body.scrollHeight - window.innerHeight);
+  }, [lastAnchorElementExists]);
+
+  // Find the href for which the element is within the viewport
+  const findHrefInView = useCallback(() => {
+    return hrefs.find(href => {
+      const rect = getRectByHref(href);
+      return rect && rect.bottom <= window.innerHeight && rect.bottom - scrollSpyOffset >= 0;
```
Comment (jperals) [inline src/anchor-navigation/scroll-spy.tsx:38]: Why is it
```rect.bottom - scrollSpyOffset >= 0``` and not
```rect.top - scrollSpyOffset >= 0```?
(or `rect.top >= scrollSpyOffset` for simplicity)

Comment (rubencarvalho) [reply src/anchor-navigation/scroll-spy.tsx]: Simplified it!

File: src/anchor-navigation/__integ__/anchor-navigation.test.ts [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+  test(
+    'the correct aria-labelledby attribute is applied',
+    setupTest(async page => {
+      return expect(await page.getElementAttribute(wrapper.toSelector(), 'aria-labelledby')).toBe('anchor-nav-heading');
+    })
+  );
+
+  test(
+    'clicking in a anchor link makes the respective anchor item active',
```
Comment (jperals) [inline src/anchor-navigation/__integ__/anchor-navigation.test.ts:41]: Minor typo: "**an** anchor link" :)

File: src/anchor-navigation/scroll-spy.tsx [hunk 16]
```diff
@@ -0,0 +1,79 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (jperals) [inline src/anchor-navigation/scroll-spy.tsx:1]: _Almost_ every file in the repo that contains only a hook has the `use-` prefix in the file name. I'd suggest to follow this pattern here as well.

File: src/anchor-navigation/scroll-spy.tsx [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+  const getRectByHref = useCallback(href => {
+    return document.getElementById(href.slice(1))?.getBoundingClientRect();
+  }, []);
+
+  // Check if we're scrolled to the bottom of the page
+  const isPageBottom = useCallback(() => {
+    return lastAnchorExists && Math.ceil(window.scrollY) >= Math.floor(document.body.scrollHeight - window.innerHeight);
+  }, [lastAnchorExists]);
+
+  // Find the href for which the element is within the viewport
```
Comment (jperals) [inline src/anchor-navigation/scroll-spy.tsx:37]: The _first_ href, right?

File: src/anchor-navigation/scroll-spy.tsx [hunk 18]
```diff
@@ -0,0 +1,79 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { useCallback, useEffect, useState } from 'react';
+
+const isBrowser = typeof window !== 'undefined';
+
+export default function useScrollSpy({
```
Comment (jperals) [inline src/anchor-navigation/scroll-spy.tsx:7]: Looks like this hook assumes that the hrefs, and by extension the anchors passed in in the props, are sorted by order of appearance in the page. Is this a safe assumption? Or should be either a) make it explicit in the API description or b) let the component sort them

Comment (rubencarvalho) [reply src/anchor-navigation/scroll-spy.tsx:7]: Good point. I'd be for `a)`.
For `b)`, I can imagine this could add some computational overhead, especially if the list of `hrefs` is long or if the DOM is complex. Also, I feel it becomes _too much_ magic.

**Review by jperals — APPROVED** [review]
